[Skia] Incorrect font rendering of composed characters with skia and certain fonts

carlosgcampos opened the pull request:
#### d00fe9e28b11e77857afbe966094a3c23c1cf306
<pre>
[Skia] Incorrect font rendering of composed characters with skia and certain fonts
<a href="https://bugs.webkit.org/show_bug.cgi?id=291272">https://bugs.webkit.org/show_bug.cgi?id=291272</a>

Reviewed by Adrian Perez de Castro.

The problem happens with some sequences like &quot;R̥g&quot;
which is composed of characters (#x0052 #x0325)
LATIN CAPITAL LETTER R with COMBINING RING BELOW. Since the system
fallback only allows to check for a single character, if we find a font
that contains a glyph for &apos;R&apos; but that can&apos;t render the combining
sequence we end up falling back to the last resort fallback font, which
depending in the system might not be able to render the sequence either.
So, when we fail to find a font that can render the sequence, we should
try to find another font in the system that contains all other
characters in the sequence as we do with cairo. But Skia API doesn&apos;t
allow to query a set of characters, we would need to call the
matchFamilyStyleCharacter() for every character which would be very
slow. We need to use fontconfig API for that, looking for a font that
contains all characters in the sequence or at least the one having more
charcter coverage. We can take advantage that we have to use fontconfig
API directly for system fallbacks to being back the FontSet cache that
will make it even faster. See <a href="https://bugs.webkit.org/show_bug.cgi?id=203544">https://bugs.webkit.org/show_bug.cgi?id=203544</a>

Canonical link: <a href="https://commits.webkit.org/298033@main">https://commits.webkit.org/298033@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a70599c28a0aaa7401206cee998e5e6376994b00

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113009 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32744 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23222 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119213 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64208 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114971 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33396 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41307 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86004 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41410 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115956 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26658 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101650 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66317 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25934 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19782 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62971 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96044 "Found 1 new API test failure: TestWebKitAPI.WKWebsiteDataStoreConfiguration.TotalQuotaRatioWithPersistedDomain (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19856 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122434 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40087 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29886 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94854 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40471 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97871 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94597 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39735 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17542 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36200 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18276 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39973 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45472 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39614 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42947 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41351 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebCore/platform/graphics/skia/SkiaSystemFallbackFontCache.cpp:
```diff
@@ -0,0 +1,319 @@
+    };
+    weight = mapRanges(style.weight(), weightRanges, std::size(weightRanges));
+
+    static constexpr MapRanges widthRanges[] = {
```
We could use `std::array<MapRanges> widthRanges` instead. It can equally be a `constexpr`, and allows us to change the method signature in the `mapRanges` lambda so it no longer takes a `rangesCount` -- which is safer, as it ensures we query the size from the container, and does not impact the rest of the logic.

carlosgcampos replied:
std::array needs the size in the template declaration, so we can't use the same lambda for the two arrays. We could use a Vector, but then it can't be constexpr

spenap replied:
I see -- and if we want the array, then we would need to use a template and not a lambda. That's fine then, sorry

spenap commented on Source/WebCore/platform/graphics/skia/SkiaSystemFallbackFontCache.h:
```diff
@@ -0,0 +1,52 @@
+    SkiaSystemFallbackFontCache();
+    ~SkiaSystemFallbackFontCache();
+
+    sk_sp<SkTypeface> fontForCharacterCluster(const SkFontStyle&, const String& locale, StringView);
```
I guess we could argue this is `const` semantically, and mark the `m_cache` below `mutable`.

carlosgcampos replied:
I don't follow, the method changes the cache, it's not const.

spenap replied:
It depends on habit or style. You can grep the WebKit code for `mutable` and it appears often with caches. In many codebases, when you look up a key, that's semantically considered `const` even if the cache is updated. But no strong opinion.

spenap commented on Source/WebCore/platform/graphics/skia/SkiaSystemFallbackFontCache.cpp:
```diff
@@ -0,0 +1,319 @@
+            FcFontSetDestroy(m_fontSet);
+        }
+
+        const FontSetCache::Font* bestForCharacterCluster(FcCharSet* charSet)
```
```suggestion
        const FontSetCache::Font* bestForCharacterCluster(FcCharSet* charSet) const
```

aperezdc commented on Source/WebCore/platform/graphics/skia/SkiaSystemFallbackFontCache.cpp:
```diff
@@ -0,0 +1,319 @@
+    public:
+        static std::unique_ptr<FontSet> create(const String& locale, int weight, int width, int slant)
+        {
+            auto* pattern = FcPatternCreate();
```
We have `Source/WebCore/platform/graphics/freetype/RefPtrFontconfig.h` which defines the needed method to use `RefPtr<FcPattern>`, so I would prefer to use a smart pointer here:

```suggestion
            RefPtr<FcPattern> pattern = adoptRef(FcPatternCreate());
```

aperezdc replied:
And we could add the needed `RefPtr` traits for other Fontconfig types. Some of the API names are misleading and actually add reference counts instead of creating new objects i.e. `FcCharSetCopy()`, and letting our smart pointers take care of the details would avoid that kind of wart.

Note that some Fontconfig types&mdash;typically the ones without a `FcFooCopy` function&mdash;are _not_ reference counted, i.e. `FcFontSet` so for those we would use `std::unique_ptr<>` instead.

carlosgcampos replied:
I thought about it, but since this is the only place where we need to use fontconfig directly I didn't want to depend on anything from platform/freetype (which is indeed confusing, this is not about freetype) in platform/Skia. There are very few manual creare/destroy so I don't think it's worth it. An alternative would be to add our own unique_ptr in this file and use it for all cases, since we don't really need to take references.

aperezdc replied:
Right, good point about avoiding depending on `platform/freetype`. I suppose we could move `RefPtrFontconfig.h` into `platform/fontconfig` and use it from different places, but let's land this patch as-is; we can do the cleanup later, _specially_ if we end up needing more code that uses Fontconfig later on.

carlosgcampos committed:
[Skia] Incorrect font rendering of composed characters with skia and certain fonts
https://bugs.webkit.org/show_bug.cgi?id=291272

Reviewed by Adrian Perez de Castro.

The problem happens with some sequences like "R̥g"
which is composed of characters (#x0052 #x0325)
LATIN CAPITAL LETTER R with COMBINING RING BELOW. Since the system
fallback only allows to check for a single character, if we find a font
that contains a glyph for 'R' but that can't render the combining
sequence we end up falling back to the last resort fallback font, which
depending in the system might not be able to render the sequence either.
So, when we fail to find a font that can render the sequence, we should
try to find another font in the system that contains all other
characters in the sequence as we do with cairo. But Skia API doesn't
allow to query a set of characters, we would need to call the
matchFamilyStyleCharacter() for every character which would be very
slow. We need to use fontconfig API for that, looking for a font that
contains all characters in the sequence or at least the one having more
charcter coverage. We can take advantage that we have to use fontconfig
API directly for system fallbacks to being back the FontSet cache that
will make it even faster. See https://bugs.webkit.org/show_bug.cgi?id=203544

Canonical link: https://commits.webkit.org/298033@main
